### ISSUE TITLE
OpenShift and management commands

### ISSUE THREAD

**Issue by alesrosina** [WeblateOrg/weblate#2677]
When Weblate is deployed on OpenShift none of management commands (https://docs.weblate.org/en/latest/admin/management.html) that use `git`, does not work. You would always get an error:

```
No user exists for uid XXXXXXXXX
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

This is solved in actual app, but it seems that management commands use some kind of different logic for accessing git and it always fails.

**Comment by alesrosina** [comment]
@GrahamDumpleton I would again need your help here, I've spent a day trying to figure it out but got completely lost. I thought `openshift/python` image, which is used to build weblate image for deployment on openshift has already `nss_wrapper` implemented, but this somehow does not work. In general - if i run `ssh whatever.example.com`, I will always get this error, even if I run it via
```
#!/bin/sh
    exec ssh \
        -o "UserKnownHostsFile=/opt/app-root/data/ssh/known_hosts" \
        -o "IdentityFile=/opt/app-root/data/ssh/id_rsa" \
        -o StrictHostKeyChecking=yes \
        -o HashKnownHosts=no \
        "$@"
```
which is generated by Weblate in `/opt/app-root/data/ssh/ssh-weblate-wrapper-[some-md5-hash]`.

What am I missing here?

**Comment by nijel** [comment]
Maybe you execute management command under different user than is actually running Weblate?

**Comment by alesrosina** [comment]
I am not sure actually what is going on here :)

That's why I mentioned Graham, since he is quite an expert on Openshift and also did the initial setup of it. I am probably missing out sth quite obvious, but right now I just cannot get my head around it.

**Comment by GrahamDumpleton** [comment]
Try using:

```
#!/bin/bash
```

for your script, not normal ``/bin/sh``.

The ``nss_wrapper`` stuff relies on a shell initialisation trick. I can't remember if it works for ``sh``, or only ``bash``.

**Comment by GrahamDumpleton** [comment]
BTW, how exactly are you executing this script inside of the container. That is probably the more important question. Is it executed from the web interface, or are you using ``oc rsh`` to execute it from outside?

**Comment by alesrosina** [comment]
Hey @GrahamDumpleton thanks for this hint.
Sadlly I did not manage to make it work, even with `#!/bin/bash`. 

This error appears on web interface, when I'm doing `oc rsh` it's even worse - I got this error:
```
python: error while loading shared libraries: libpython3.5m.so.rh-python35-1.0: cannot open shared object file: No such file or directory
command terminated with exit code 127
```

Thanks again for your help.

**Comment by GrahamDumpleton** [comment]
What is the ``oc rsh`` command you are running? Are you doing that to get a shell and then invoking the command, or passing the command as argument to ``oc rsh``?

If using ``oc rsh`` to get a shell, what do you get when you run from that shell:

```
env | grep NSS
```

and:

```
env | grep LD_LIBRARY_PATH
```

**Comment by alesrosina** [comment]
Interesting, those two suggestions are returning nothing when using `oc rsh`, but when using web console, first one returns nothing, second one returns 
```
LD_LIBRARY_PATH=/opt/rh/rh-python35/root/usr/lib64:/opt/rh/rh-nodejs8/root/usr/lib64:/opt/rh/httpd24/root/usr/lib64
```

**Comment by GrahamDumpleton** [comment]
Does Weblate work with Python 3.6.

What you are probably hitting is:

* https://bugzilla.redhat.com/show_bug.cgi?id=1430327

There is even an existing workaround for this bug in:

* https://github.com/WeblateOrg/weblate/blob/master/.s2i/action_hooks/deploy#L30

You appear to be the one who added that workaround. I didn't know that problem existed.

**Comment by alesrosina** [comment]
Yes, I did add this, but it did not solve this issue actually.

Do you think increasing a python version to 3.6 would fix this?

**Comment by GrahamDumpleton** [comment]
The setting of the ``LD_LIBRARY_PATH`` for the workaround should really have been done in ``deploy_env`` and not ``deploy``. Any variables set in ``deploy_env`` will be automatically exported in context of ``deploy`` script.

* https://github.com/WeblateOrg/weblate/blob/master/.s2i/action_hooks/deploy_env

If you do that and then run:

```
powershift image shell
```

it will set up your shell environment and also inherit anything set in ``deploy_env``.

The ``powershift image shell`` command is therefore the way you can get a shell with the same environment settings as the running web application itself got.

So move that workaround into ``deploy_env`` and then run ``powershift image shell`` to get same environment as application does, before running any commands.

Separate to above, my issue is that the repo will not even install on OpenShift anymore. Issues have found:

* For sqlite deployment, the sqlite database version in image is too old and Django dies.

* For postgresql deployment, database is okay, except that the secrets you added to the templates are missing the labels that all other resources has, which means when you delete the application resources using a label selector, it leaves the secrets behind and they have to be deleted separately. If you don't delete the secrets and create from the template again, something mucks up as passwords out of sync somewhere, possibly because the template params are passed in direct as environment variables in a deployment rather than being used from the secret.

* When postgresql deployment comes up, it fails because there is no ``ssh`` directory in PVC.

```
 -----> Set chmod 600 for id_rsa keys.
chmod: cannot access ‘/opt/app-root/data/ssh/id_rsa’: No such file or directory 
```

**Comment by alesrosina** [comment]
Thanks for this. I will move this to `deploy_env`.

Regarding issues:
- should we drop sqlite aynway? I don't see a benefit of using it on openshift at all.
- The problem here was, that on build script, in `settings_openshift.py` (see https://github.com/WeblateOrg/weblate/blob/master/weblate/settings_openshift.py#L35) will request sqlite, and if we configured it to use postgres, sqlite should not be used at build step. Not sure if there is any other option to skip database setup in build step?
- you're right, i have to do a check if file actually exists before. This is added for a problem, when this file is added via weblate and you do a redeploy of application - it needs to reset file permissions, since by default openshift will change permissions (i guess) when attaching a storage to image.

I have opened an MR with fixes discussed here: https://github.com/WeblateOrg/weblate/pull/2678
